File: app/src/main/java/com/example/fantastiqa/GameState/Board.java
package com.example.fantastiqa.GameState;

import android.util.Pair;
import com.google.common.graph.MutableNetwork;
import com.google.common.graph.NetworkBuilder;
import com.google.common.graph.ElementOrder;
import com.google.common.graph.Network;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;
import java.util.Iterator;
import java.util.Set;

public class Board {
    public final List<Quest> quests = new ArrayList<>(2);
    public MutableNetwork<Region,Road> regionsRoads= NetworkBuilder.undirected().expectedEdgeCount(7).expectedNodeCount(6).build();;

    public Board() {
        Random  r = new Random();

        List<RegionName> regions = new ArrayList<>(Arrays.asList(RegionName.values()));
        List<TowerName> towers = new ArrayList<>(Arrays.asList(TowerName.values()));
        towers.addAll(Arrays.asList(TowerName.values()));

		while (regions.size()>0) {
			regionsRoads.addNode(new Region(regions.remove(r.nextInt(regions.size())),
				towers.remove(r.nextInt(towers.size()))));
		}
		
		Iterator<Region> regionsIter = regionsRoads.nodes().iterator();
		int regionCount = 0;
		Region regionMiddle1 = null;
		Region regionFirst = (regionsIter.hasNext() ? regionsIter.next() : null);
		Region regionVeryFirst = regionFirst;
		Region regionSecond = null;
		while (regionsIter.hasNext()) {
			regionSecond = regionsIter.next();
			if (regionCount == 1) {
				regionMiddle1 = regionFirst;
			}
			if (regionCount == 4) {
				regionsRoads.addEdge(regionMiddle1, regionFirst, new Road());
			}
			regionsRoads.addEdge(regionFirst, regionSecond, new Road());
			regionFirst = regionSecond;
			regionCount++;
		}
		regionsRoads.addEdge(regionFirst, regionVeryFirst, new Road());
    }

    public List<Pair<Road, Region>> getAdjacentAreas(Region starting) {
        List<Pair<Road,Region>> result = new LinkedList<>();
        for (Region adjRegion : regionsRoads.adjacentNodes(starting)) {
			result.add(new Pair<>(regionsRoads.edgeConnectingOrNull(starting, adjRegion), adjRegion));
		}
		return result;
    }

    public List<Region> regions(){
		return new ArrayList<>(regionsRoads.nodes());
    }

    public List<Road> roads() {
		return new ArrayList<>(regionsRoads.edges());
    }
    
    public Region getTowerMatch(Region startRegion) {
		for (Region aRegion : regionsRoads.nodes()) {
			if (aRegion != startRegion && aRegion.tower == startRegion.tower) {
				return aRegion;
			}
		}
		return null;
	}
	
	public Road getRoad(Region r1, Region r2) {
		return regionsRoads.edgeConnectingOrNull(r1,r2);
	}
	
}

File: app/src/main/java/com/example/fantastiqa/gameState/Board.kt
package com.example.fantastiqa.gameState

import com.example.fantastiqa.pieces.RegionName
import com.example.fantastiqa.pieces.TowerName
import java.util.*

/**
 * Immutable Board representation.
 */
data class Board(
    @JvmField val quests: List<BoardQuest?> = listOf(null, null),
    @JvmField val adjacencies: Map<Region, Map<Region, Road>> = emptyMap()
) {

    /**
     * Returns a new board with the quest at [index] replaced.
     */
    fun withQuest(index: Int, newQuest: BoardQuest?): Board {
        val newQuests = quests.toMutableList()
        newQuests[index] = newQuest
        return copy(quests = newQuests)
    }

    /**
     * Returns a new board with the road between [r1] and [r2] replaced.
     */
    fun withRoad(r1: Region, r2: Region, newRoad: Road): Board {
        val newAdjacencies = adjacencies.toMutableMap()
        
        val r1Map = newAdjacencies[r1]?.toMutableMap() ?: mutableMapOf()
        r1Map[r2] = newRoad
        newAdjacencies[r1] = r1Map

        val r2Map = newAdjacencies[r2]?.toMutableMap() ?: mutableMapOf()
        r2Map[r1] = newRoad
        newAdjacencies[r2] = r2Map

        return copy(adjacencies = newAdjacencies)
    }

    fun getAdjacentAreas(starting: Region): List<Pair<Road, Region>> {
        return adjacencies[starting]?.map { (adjRegion, road) ->
            Pair(road, adjRegion)
        } ?: emptyList()
    }

    fun regions(): List<Region> = adjacencies.keys.toList()

    fun roads(): List<Road> {
        val identitySet = Collections.newSetFromMap(IdentityHashMap<Road, Boolean>())
        adjacencies.values.forEach { innerMap ->
            identitySet.addAll(innerMap.values)
        }
        return identitySet.toList()
    }

    fun getTowerMatch(startRegion: Region): Region? {
        return adjacencies.keys.find { it != startRegion && it.tower == startRegion.tower }
    }

    fun getRoad(r1: Region, r2: Region): Road? = adjacencies[r1]?.get(r2)

    companion object {
        /**
         * Factory method to create an initial board.
         */
        @JvmStatic
        fun createInitialBoard(): Board {
            val r = Random()
            val regionsNames = RegionName.values().toMutableList()
            val towers = (TowerName.values().toList() + TowerName.values().toList()).toMutableList()

            val regionNodes = mutableListOf<Region>()
            while (regionsNames.isNotEmpty()) {
                val region = Region(
                    regionsNames.removeAt(r.nextInt(regionsNames.size)),
                    towers.removeAt(r.nextInt(towers.size))
                )
                regionNodes.add(region)
            }

            var tempBoard = Board(adjacencies = regionNodes.associateWith { emptyMap<Region, Road>() })

            if (regionNodes.isNotEmpty()) {
                var regionMiddle1: Region? = null
                var regionFirst = regionNodes[0]
                val regionVeryFirst = regionFirst

                for (i in 1 until regionNodes.size) {
                    val regionSecond = regionNodes[i]
                    if (i == 2) regionMiddle1 = regionNodes[1]
                    
                    if (i == 5 && regionMiddle1 != null) {
                        tempBoard = tempBoard.addRoad(regionMiddle1, regionFirst, Road())
                    }
                    
                    tempBoard = tempBoard.addRoad(regionFirst, regionSecond, Road())
                    regionFirst = regionSecond
                }
                tempBoard = tempBoard.addRoad(regionFirst, regionVeryFirst, Road())
            }
            return tempBoard
        }

        private fun Board.addRoad(r1: Region, r2: Region, road: Road): Board {
            val newAdjacencies = adjacencies.toMutableMap()
            
            val r1Map = newAdjacencies[r1]?.toMutableMap() ?: mutableMapOf()
            r1Map[r2] = road
            newAdjacencies[r1] = r1Map

            val r2Map = newAdjacencies[r2]?.toMutableMap() ?: mutableMapOf()
            r2Map[r1] = road
            newAdjacencies[r2] = r2Map

            return copy(adjacencies = newAdjacencies)
        }
    }
}
